defaults/main.yml

```diff
@@ -295,6 +298,50 @@ galaxy_app_config_default:
   # Everything else
   visualization_plugins_directory: "config/plugins/visualizations"
 
+  # Static and themes configuration, will only be added if galaxy_manage_themes
+  enable_static: "{{ galaxy_manage_subdomain_static }}"
+  static_dir_by_host: >
+    { {% if galaxy_manage_subdomain_static %}
+    {% for subdomain in galaxy_themes_subdomains %}
+    '{{ subdomain.name }}.{{ galaxy_themes_instance_domain}}': '{{ galaxy_themes_static_path }}/static-{{ subdomain.name }}/'
+    {% endfor %}
+    {% endif %} }
```

Comment:
I'm looking at alternative implementations here for what we can do.

I was thinking perhaps a solution here is to separate these out and leverage ansible's resolution process for variables. Here we could write the following (we should probably do it anyway just for cleanliness but).

```suggestion
  static_dir_by_host: {{ __static_dir_by_host }}
```

but ansible's `set_fact` module considers variable names with a `.` as invalid, they're literally forbidden whcih is wild to me.

```
---
- hosts: all
  vars:
    galaxy_themes_static_keys:
      static_dir: ""
      static_images_dir: "images/"
      static_scripts_dir: "scripts/"
      static_welcome_html: "welcome.html"
      static_favicon_dir: "favicon.ico"
      static_robots_txt: "robots.txt"
  pre_tasks:
    - find:
        paths: files/galaxy/static/
        patterns: "{{ galaxy_themes_static_keys.static_welcome_html }}"
        recurse: yes
      register: python_files

    - debug:
        msg: "{{ item.path }}"
      with_items: "{{ python_files.files }}"

    - set_fact:
        "{{ item.path | regex_replace('files/galaxy/static/', '') | regex_replace('/static/.*', '')}}": "{{ item.path }}"

      with_items: "{{ python_files.files }}"
      # with_fileglob: 'files/galaxy/static/*/static/{{ galaxy_themes_static_keys.static_welcome_html }}'


    - debug:
        msg: "{{ __static_dir_by_host }}"
```

so that's out. I tried a couple different ways including `from_yaml` to accomplish this, but, none worked.

It would be a *lot* easier if we could pull this information from another file, rather than sticking it all in galaxy.yml. @natefoo is there any good way to accomplish that? the only other option is appending to the end of galaxy.yml manually which is ,,, no. I don't want that.

Reply:
OK it works with from_json, ALRIGHT.

Reply:
<details><summary>Ignore this version I was dumb</summary>

```
    - find:                                                               
        paths: files/galaxy/static/                                       
        patterns: "{{ galaxy_themes_static_keys.static_welcome_html }}"   
        recurse: yes                                                      
      register: python_files
     # This is alternatively just `galaxy_themes_subdomains`, but I did dynamic discovery for some dumb reas.

    - set_fact:                                                                                                         
        __static_dir_by_host: |                                                                                         
          {{ '{' }}                                                                                                     
          {% for item in python_files.files %}                                                                          
            "{{ item.path | regex_replace('files/galaxy/static/', '') | regex_replace('/static/.*', '')}}": "{{ item.path }}",
          {% endfor %}                                                                                                  
          {{ '}' }}
    - debug:                                                   
        msg: "{{ __static_dir_by_host }}"                      
                                                               
    - debug:                                                   
        msg: "{{ __static_dir_by_host['test.example.com'] }}"                                  
```
</details>


yeah this works:

```
    - set_fact:                                                                                                         
        __static_dir_by_host: |                                                                                         
          {{ '{' }}                                                                                                     
          {% for subdomain in galaxy_themes_subdomains %}                                                                          
            "{{ subdomain.name }}.{{ galaxy_themes_instance_domain}}": "{{ galaxy_themes_static_path }}/static-{{ subdomain.name }}/images",
          {% endfor %}                                                                                                  
          {{ '}' }}
    - debug:                                                   
        msg: "{{ __static_dir_by_host }}"                      
                                                               
    - debug:                                                   
        msg: "{{ __static_dir_by_host['test.example.com'] }}"                                  
```

I'm not sure I understand why that works but it works pretty OK I think.

We should write a `tasks/subdomain_discovery.yml` that sets these `__` variables ahead of computation (we should set defaults of `__static_dir_by_host: {}`, and then override them.), a set of tasks only included conditionally on `galaxy_manage_subdomains`, and then whenever galaxy.yml gets around to being read, it'll read those variables correctly.

Reply:
Thank you so much for finding a solution :heart: 
Even though I don't understand why this works, either.
So we have to do this for each static_x_by_host thing, right? (images, robots, etc.)

Reply:
yep, one for each! Then we should be .. ready to go? exciting!

Reply:
I can add this today, but there is one more thing, unfortunately – the static_dir copying task is not idempotent unfortunately. It is done each time you run the role. Maybe we could solve this by using the rsync module?
Problem there is that it takes the same directory name, which is a problem if you want to "clone" the directory in the same spot